### PR TITLE
Flatten the panel's copper moment within a bounded per-layer trade

### PR DESCRIPTION
**Pull request by akhilles**
Copper balancing preserved whatever through-stack imbalance the boards were drawn with. It measured the moment of each layer's *deviation from its target*, which is zero exactly when every layer lands on target — the normal case since #1024 — so a correctly balanced panel reported nothing left to flatten while still carrying real asymmetry.

This reads the density itself, so the moment is of the copper actually present, and gives each layer a bounded budget to answer it with.

## The metric

```
M_i = sum_l w_l rho_li          (was: sum_l w_l (rho_li - d_l))
```

Penalizing this field pointwise covers both warp modes with one term rather than separate machinery: **bow is the field's mean, twist is its variation**, and squaring it at every site charges for both.

## The budget

The spatial solve used to conserve each layer's copper area exactly, so it could redistribute within a layer but never trade one layer's density match for a flatter moment. Each layer now has a band around its selected area, and the bands are settled jointly so the stack's density deviations cancel — a layer may only take on what another gives back.

Two settings, separate jobs:

- `stack_moment_weight` (10) scales the moment against the local density match. Equal weighting is far too timid: the local term brings one squared error per layer against this one's single term, so a six-layer stack sheds only about a sixth of a uniform moment.
- `stack_flex_density` (1%) caps where any layer can end up, whatever the weight asks for. It is what refuses a two-layer panel poured 0.9 against 0.3 the 30% correction it wants.

Zero flex reproduces the pinned result exactly; a stackup with no weights is unaffected by either.

## Why the bands are coupled

Freeing each layer's total independently frees a direction the stack cannot see. A symmetric stackup's weights sum to zero, so every layer drifting the same way leaves the moment untouched — and the local term will spend the whole band on exactly that drift, since bare clearance reads as a deficit on every layer at once.

Measured with independent bands: all six layers rose by the full 1%, no mirror gap moved, and the two densest layers were pushed into the corner of the radius box, flattening their voids to a single radius. Coupling removes that failure mode.

One linear equality over separable convex sets dualizes to a single multiplier, so each layer projects its own commonly shifted proposal onto its box and band, with the multiplier found by bisection. A layer pinned at a band edge stops responding to it, so a trial costs one clamped sum per layer rather than a nested projection.

## Effect

6-layer A6 array, normalized stack moment **0.0018 -> 0.0012**:

| Layer | Target | Before | After | Δ |
|---|---|---|---|---|
| F.Cu | 0.784 | 0.784 | 0.788 | +0.004 |
| In1.Cu | 0.923 | 0.923 | 0.916 | −0.007 |
| In2.Cu | 0.885 | 0.885 | 0.886 | +0.001 |
| In3.Cu | 0.877 | 0.877 | 0.879 | +0.002 |
| In4.Cu | 0.923 | 0.923 | 0.916 | −0.007 |
| B.Cu | 0.779 | 0.779 | 0.785 | +0.006 |

F.Cu and B.Cu close from 0.005 apart to 0.003, no layer ends more than 0.007 off its target, and the trade nets to −0.001. The five-up 18x24 panel built from those arrays falls **0.0011 -> 0.0005** — the improvement compounds, since the panel inherits already-flattened arrays. Both numbers are printed in the balance summary, so the trade is never silent.

Panel generation goes 24.4s -> 31.0s wall at unchanged total CPU; the coupled projection is one serial scalar solve per iteration, with the per-layer projections fanned back out.

`docs/board-array-copper-balancing.md` is rewritten where it required the opposite behavior.

---

> [!NOTE]
> **Medium Risk**
> Changes core panelization geometry and numerical optimization defaults; bounded by profile knobs and extensive tests, but affects fabrication copper fill on all balanced panels.
> 
> **Overview**
> Copper balancing now targets the **panel’s actual through-stack copper moment** (weighted density about the mid-plane), not per-layer deviation from target—so asymmetric boards that already hit their targets can still be flattened.
> 
> The spatial solver adds **`stack_moment_weight`** and **`stack_flex_density`** (defaults 2.0 and 1%): layers may shift density within a band so copper trades between layers under a **joint equality** that keeps stack deviations net-zero. Projection moves from pinned per-layer sums to **`project_coupled_bands`** (bisection on a shared multiplier plus per-layer box/band projection). **`stack_flex_density: 0`** restores the old pinned behavior; missing stack weights keep bands closed.
> 
> **`generate_spatial_dense_copper_balance`** now returns **`SpatialCopperBalance`** (layer results plus optional **`StackMomentField`** RMS before/after). IPC tooling reports stack moment and field RMS in **`CopperBalanceReport::summary_lines`**. Docs and changelog describe the new metric and settings; tests cover moment shrinkage, flex caps, coupled bands, and edge cases (no lattice, no weights).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 573d68f5657470727a03ae48ae5bb5a437339576. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: crates/pcb-ir/src/geom/copper_balance/mod.rs [hunk 1]
```diff
@@ -1484,6 +1524,143 @@ mod tests {
...
+        // Neither leaves its band, and the stack gains nothing overall.
+        for deviation in &deviations {
+            assert!(deviation.abs() <= flex + quantization, "{deviations:?}");
+        }
+        assert!(
+            deviations.iter().sum::<f64>().abs() <= 2.0 * quantization,
+            "{deviations:?}"
+        );
+    }
+
```
[the remark this answers is not in the document]
Comment (akhilles) [reply crates/pcb-ir/src/geom/copper_balance/mod.rs:2003]: Valid — fixed in 79fa913a (rebased).

`spatial_solver_uses_signed_stack_weights` now pins `stack_flex_density: 0.0` explicitly. Its subject is the redistribution the stack weights drive *at a fixed copper area*, which is worth testing in isolation, so pinning states that rather than inheriting whatever `V1` happens to default to.

Checked the sibling area-preservation tests too — `spatial_solver_preserves_minimum_radius_area`, `..._preserves_area_and_radius_bounds_for_constant_inputs` and `..._opposes_a_fixed_copper_gradient_without_changing_total_area` are all single-layer, and with one layer the zero-sum constraint has no trade partner and pins the sum exactly. Their invariants still hold by construction. This was the only test that could actually trade.

File: crates/pcb-ir/src/geom/copper_balance/spatial.rs [hunk 2]
```diff
@@ -271,33 +271,143 @@ fn scanline_indicator(points: &[Point], region: &ContourSet) -> Vec<f64> {
...
+                .collect::<Vec<_>>();
+            project_box_interval(
+                &shifted,
+                lower,
+                upper,
+                layer.sum_bounds.0,
+                layer.sum_bounds.1,
+            )
+        })
+        .collect()
```
[the remark this answers is not in the document]
Comment (akhilles) [reply crates/pcb-ir/src/geom/copper_balance/spatial.rs]: Half right, and the half that is right is worth fixing — done in 6a203351.

Measured on an 18x24 five-up panel (80811 sites x 6 layers), before vs after this PR:

| | user CPU | wall | util |
|---|---|---|---|
| per-layer projection | 83.6s | 24.4s | 350% |
| coupled (as reviewed) | 83.2s | 33.3s | 253% |
| coupled + this fix | 77.6s | 26.9s | 293% |

Total CPU was unchanged at 83s, so "an order of magnitude more work" and "many times longer" are not what happens — the multiplier search is 64 passes of a clamp-and-sum, which is cheap next to the convolution that dominates the loop. What was real is the serialization: utilization fell from 350% to 253% and the panel lost a third of its wall clock.

Fixed by fanning the final projections back out over the same scoped threads the rest of the loop uses, and stopping the multiplier search once the constraint holds to a density finer than the radius quantization can express. Geometry is identical.

Skipped the degenerate-band short-circuit: `V1` ships a non-zero flex, so it would be a branch that never fires on the default path.

File: crates/pcb-ir/src/geom/copper_balance/mod.rs [hunk 3]
```diff
@@ -498,39 +522,45 @@ pub fn generate_spatial_dense_copper_balance(
...
+                / ROUNDED_HEXAGON_AREA_FACTOR;
+            // A layer that took no lattice has no radii to trade with, and its
+            // sites are not its own to spend: sizing the box range from them
+            // would put the band above a sum that is fixed at zero.
+            let site_count = squared_radii[layer_index].len() as f64;
+            (
+                (sum - slack).max(site_count * lower),
+                (sum + slack).min(site_count * upper),
+            )
         })
```
[the remark this answers is not in the document]
Comment (akhilles) [reply crates/pcb-ir/src/geom/copper_balance/mod.rs:635]: Valid, and thank you — but not for the reason given, and I nearly dismissed it on that basis.

The mechanism described cannot invert the band. `site_count` comes from the layer's own radii, and every radius from the uniform solve lies in `[r_min^2, r_max^2]`, so `sum` always lies between `site_count * lower` and `site_count * upper` and the clamps cannot cross. That path was fixed in ac8e1f8b; the cited lines 558-561 predate it.

The inversion is real anyway, one level down. `sum` is accumulated one radius at a time while the bounds are products, so a layer whose radii all pin to one bound rounds across it. Reproduced in 055c03bd: `min 136.89000000000013` against `max 136.89000000000001`, panicking the clamp.

It needed the flex band shut to surface, which is why it appeared while fixing your other finding rather than before. Both ends are now anchored on the sum itself, which is what the band is required to contain, and `a_saturated_radius_field_still_yields_a_usable_band` fails without that anchor.

File: crates/pcb-ir/src/geom/copper_balance/mod.rs [hunk 4]
```diff
@@ -498,42 +569,53 @@ pub fn generate_spatial_dense_copper_balance(
...
-        .collect::<Vec<Vec<f64>>>();
+        .collect::<Vec<_>>();
```
[the remark this answers is not in the document]
Comment (akhilles) [reply crates/pcb-ir/src/geom/copper_balance/mod.rs:636]: Valid — fixed in 055c03bd.

You are right that the band opened regardless of whether any weight existed, and right about the consequence: with all weights zero there is no moment term, so the only thing left to spend the band on is the local density error, which trades layers off their targets for nothing. That directly contradicted the summary line telling the user those layers were balanced independently, and my own docstring claiming a weightless stackup lands in the same place either way — true of `stack_moment_weight`, false of `stack_flex_density`.

The slack is now zero unless some weight is non-zero, so a panel without stackup thickness data holds every layer on its target. `layers_hold_their_targets_when_the_stackup_weighs_nothing` pins it.

Closing the band also surfaced a latent rounding inversion underneath it — see the reply on your other thread.

File: CHANGELOG.md [hunk 5]
```diff
@@ -11,6 +11,7 @@ and this project adheres to Semantic Versioning (https://semver.org/spec/v2.0.0.
 ### Changed
 
 - `--layout-target` now takes `board` or `board-array` on every command, and defaults to `board-array`.
+- Copper balancing can shift a layer up to 1% off its target density to counterweight the rest of the stack, paid for by an opposite shift elsewhere.
```
[the remark this answers is not in the document]
Comment (akhilles) [reply CHANGELOG.md:15]: Valid — fixed in 055c03bd. Only the band had an entry; the two other user-visible changes now do as well:

- balancing flattens the copper the panel actually carries about its mid-plane, rather than preserving the imbalance the boards were drawn with
- the balance summary reports the panel stack moment before and after
